client/windows/settings/SettingsMainWindow.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * SettingsMainContainer.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+
+#include "SettingsMainWindow.h"
+
+#include "GeneralOptionsTab.h"
+#include "AdventureOptionsTab.h"
+#include "BattleOptionsTab.h"
+#include "OtherOptionsTab.h"
+
+#include "filesystem/ResourceID.h"
+#include "CGeneralTextHandler.h"
+#include "gui/CGuiHandler.h"
+#include "lobby/CSavingScreen.h"
+#include "widgets/Buttons.h"
+#include "widgets/Images.h"
+#include "widgets/ObjectLists.h"
+#include "CGameInfo.h"
+#include "CPlayerInterface.h"
+#include "CServerHandler.h"
+
+
+SettingsMainWindow::SettingsMainWindow(BattleInterface * parentBattleUi) : InterfaceObjectConfigurable()
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+
+	const JsonNode config(ResourceID("config/widgets/settings/settingsMainContainer.json"));
+	addCallback("activateSettingsTab", [this](int tabId) { openTab(tabId); });
+	addCallback("loadGame", [this](int) { loadGameButtonCallback(); });
+	addCallback("saveGame", [this](int) { saveGameButtonCallback(); });
+	addCallback("restartGame", [this](int) { restartGameButtonCallback(); });
+	addCallback("quitGame", [this](int) { quitGameButtonCallback(); });
+	addCallback("returnToMainMenu", [this](int) { mainMenuButtonCallback(); });
+	addCallback("closeWindow", [this](int) { backButtonCallback(); });
+	build(config);
+
+	std::shared_ptr<CPicture> background = widget<CPicture>("background");
+	pos.w = background->pos.w;
+	pos.h = background->pos.h;
+	pos = center();
+
+	std::shared_ptr<CButton> loadButton = widget<CButton>("loadButton");
+	assert(loadButton);
+
+	std::shared_ptr<CButton> saveButton = widget<CButton>("saveButton");
+	assert(saveButton);
+
+	std::shared_ptr<CButton> restartButton = widget<CButton>("restartButton");
+	assert(restartButton);
+
+	if(CSH->isGuest())
+	{
+		loadButton->block(true);
+		saveButton->block(true);
+		restartButton->block(true);
+	}
+
+	int defaultTabIndex = 0;
+	if(parentBattleUi != nullptr)
+		defaultTabIndex = 2;
+	else if(settings["general"]["lastSettingsTab"].isNumber())
+		defaultTabIndex = settings["general"]["lastSettingsTab"].Integer();
+
+	parentBattleInterface = parentBattleUi;
+	tabContentArea = std::make_shared<CTabbedInt>(std::bind(&SettingsMainWindow::createTab, this, _1), Point(0, 40), defaultTabIndex);
```

Comment:
Don't see any point in hardcoding tab position.
If goal is to ensure that tabs buttons won't overlap with tabs content then this can be achieved in json.
And without hardcoded position it might be possible to (for example) move tab buttons into another location purely via json edit.